src/controller/PingClient.cpp

```diff
@@ -0,0 +1,146 @@
+{
+    uint64_t respTime    = static_cast<uint64_t>(System::Timer::GetCurrentEpoch());
+    uint64_t transitTime = respTime - mLastEchoTimeMillis;
+
+    mWaitingForEchoResp = false;
+    mEchoRespCount++;
+
+    ChipLogProgress(Controller, "Echo Response: %" PRIu16 "/%" PRIu16 "(%.2f%%) len=%u time=%.3fms\n", mEchoRespCount,
+                    mEchoReqCount, static_cast<double>(mEchoRespCount) * 100 / mEchoReqCount, payload->DataLength(),
+                    static_cast<double>(transitTime) / 1000);
+}
+
+CHIP_ERROR PingClient::SendEchoRequest(NodeId remoteDeviceId)
+{
+    CHIP_ERROR err = CHIP_NO_ERROR;
+
+    Messaging::SendFlags sendFlags;
+    System::PacketBufferHandle payloadBuf;
+
+    payloadBuf = MessagePacketBuffer::New(mEchoReqSize);
+    VerifyOrExit(!payloadBuf.IsNull(), err = CHIP_ERROR_NO_MEMORY);
+
+    memset(payloadBuf->Start(), 0, mEchoReqSize);
+    payloadBuf->SetDataLength(mEchoReqSize);
+
+    if (mUsingMRP)
+    {
+        sendFlags.Set(Messaging::SendMessageFlags::kNone);
+    }
+    else
+    {
+        sendFlags.Set(Messaging::SendMessageFlags::kNoAutoRequestAck);
+    }
+
+    mLastEchoTimeMillis = System::Timer::GetCurrentEpoch();
+
+    ChipLogDetail(Controller, "Send echo request message with payload size: %d bytes to Node: 0x" ChipLogFormatX64, mEchoReqSize,
+                  ChipLogValueX64(remoteDeviceId));
+
+    mWaitingForEchoResp = true;
+    err                 = mEchoClient.SendEchoRequest(std::move(payloadBuf), sendFlags);
+
+    if (err == CHIP_NO_ERROR)
+    {
+        mEchoReqCount++;
+    }
+    else
+    {
+        mWaitingForEchoResp = false;
+    }
+
+exit:
+    return err;
+}
+
+CHIP_ERROR PingClient::PingDevice(NodeId remoteDeviceId, uint16_t maxCount, Messaging::ExchangeManager * exchangeMgr)
+{
+    CHIP_ERROR err = CHIP_NO_ERROR;
+
+    VerifyOrReturnError(exchangeMgr != nullptr, CHIP_ERROR_INVALID_ARGUMENT);
+
+    // TODO: temprary create a SecureSessionHandle from node id to unblock end-to-end test. Complete solution is tracked in
+    // issue:4451
+    ReturnErrorOnFailure(mEchoClient.Init(exchangeMgr, { remoteDeviceId, 0, 0 }));
+
+    // Arrange to get a callback whenever an Echo Response is received.
+    mEchoClient.SetDelegate(this);
+
+    // Connection has been established. Now send the EchoRequests.
+    for (unsigned int i = 0; i < maxCount; i++)
+    {
+        err = SendEchoRequest(remoteDeviceId);
+
+        if (err != CHIP_NO_ERROR)
+        {
+            ChipLogError(Controller, "Send request failed: %s", ErrorStr(err));
+            break;
+        }
+
+        // Wait for response until the Echo interval.
+        while (!EchoIntervalExpired())
+        {
+            // We can use condition_varible to suspend the current thread and wake it up when response arrive after
+            // condition_varible are supported on all embedded platforms.
+            sleep(1);
```

Comment:
Thanks for including this comment.

It does feel odd to have blocking code in `src/controller`.  I would be less concerned about such a thing in `./examples`.

Reply:
Acked, made ping device operation asynchronously